### PR TITLE
Parse function2prelude

### PR DESCRIPTION
**Pull request by ekinimo**

### CODE REVIEW

File: src/infrastructure/repository.rs [hunk 1]
```diff
@@ -32,16 +32,34 @@ impl Environment<AbstractSyntaxTree> for HashMapExpressionRepository {
...
         expression_id
     }
 
-    fn get(&self, expression_id: &ExpressionId) -> Option<&AbstractSyntaxTree> {
-        self.storage.get(expression_id)
+    fn get(&self, expression_id: &ExpressionId) -> Option<AbstractSyntaxTree> {
+        match self.storage.last().unwrap().get(expression_id) {
+            Some(expr) => Some(expr.clone()),
+            None => None,
+        }
```
Comment (ekinimo) [inline src/infrastructure/repository.rs:45]: this should change. it should look at the previous hashmaps as well

File: src/domain/expression.rs [hunk 2]
```diff
@@ -4,10 +4,14 @@ pub trait Expression: Add + Mul + Sub + Div + Neg + Sized {
     type Output;
 
     fn exponentiate_by(self, power: Self) -> <Self as Expression>::Output;
+    fn make_tuple(self, power: Self) -> <Self as Expression>::Output;
+    fn make_assignment(self, power: Self) -> <Self as Expression>::Output;
+    fn make_funcall(self, power: Self) -> <Self as Expression>::Output;
```
Comment (pubrrr) [inline src/domain/expression.rs:9]: Maybe call them `assign`, `create_function_with_arguments` or something similar?

- [ ] check

File: src/infrastructure/repository.rs [hunk 3]
```diff
@@ -32,16 +32,34 @@ impl Environment<AbstractSyntaxTree> for HashMapExpressionRepository {
...
     }
 
     fn get_ids(&self) -> Vec<&ExpressionId> {
-        self.storage.keys().collect()
+        let mut vec_of_ids: Vec<&ExpressionId> = vec![];
```
Comment (pubrrr) [inline src/infrastructure/repository.rs:49]: `expression_ids`?

- [ ] rename?

File: src/infrastructure/parser.rs [hunk 4]
```diff
@@ -72,46 +74,59 @@ impl SplitStringParser {
...
+            tokens.pop();
+            let ret = Self::parse_tuple_aux(tokens)?;
+
+            match tokens.last() {
+                None => return Err("Parse error no more tokens".to_string()),
+                Some(_) => (),
+            };        
+
+            if tokens.last().unwrap() != &Token::RightParenthesis {
+                return Err("Parser error forgot to close the parenthesis".to_string());
```
Comment (pubrrr) [inline src/infrastructure/parser.rs:100]: It reads like "The parser error forgot to close the parenthesis". But usually the user forgot to do that, not the error :D

File: src/infrastructure/parser.rs [hunk 5]
```diff
@@ -72,46 +74,59 @@ impl SplitStringParser {
...
-                if !tokens.is_empty() {
-                    Err(format!(
-                        "Parser error didn't consume entire input, {:?} was left",
-                        tokens
-                    ))
-                } else {
-                    result
-                }
+            tokens.pop();
+            match &mut tokens.last() {
```
Comment (pubrrr) [inline src/infrastructure/parser.rs:103]: What does this whole thing do? Maybe extract to a function and give it a meaningful name?

- [ ] check

File: src/infrastructure/parser.rs [hunk 6]
```diff
@@ -147,11 +162,32 @@ impl SplitStringParser {
...
+                let res = AbstractSyntaxTree::variable(name.into());
+                match tokens.last() {
+                    Some(Token::LeftParenthesis) => {
+                        let tuple = Self::parse_tuple(tokens)?;
+                        Ok(res.make_funcall(tuple))
+                    }
+                    _ => Ok(res),
+                }
+            }
+            _ => {
```
Comment (pubrrr) [inline src/infrastructure/parser.rs:182]: Do we have to handle None here?

- [ ] check

File: src/infrastructure/parser.rs [hunk 7]
```diff
@@ -199,482 +251,515 @@ impl SplitStringParser {
...
-        let result = SplitStringParser::parse(tokens);
-
-        let expected = todo!();
-        // let expected = Ok(AbstractSyntaxTree::Assignment(
-        //     Assignable::Variable(Variable::from("a")),
-        //     AbstractSyntaxTree::integer(1),
-        // ));
-        // assert_eq!(result, expected);
-    }
+use parser::ExpressionParser;
```
Comment (pubrrr) [inline src/infrastructure/parser.rs:254]: fix format

- [ ] to do

File: src/infrastructure/collapsed_expression.rs [hunk 8]
```diff
@@ -130,9 +178,101 @@ impl Evaluation<CollapsedExpression<AbstractSyntaxTree>> for Node<'_> {
...
+                let (name, expr) = (children.get(0).unwrap(), children.get(1).unwrap());
+                //println!("DEBUUUUUG     name {:?}    expression  {:?}",name,expr);
+                match (name, expr) {
+                    (Node::Branch(op), node @ _) if op.operator() == Operator::FunctionCall => {
+                        let tree = match node.evaluate(&mut HashMapExpressionRepository::default())
+                        {
+                            Value(val) => AbstractSyntaxTree::float(val),
+                            Uncollapsable(tree) => tree,
+                        };
+                        let childs = op.children();
```
Comment (pubrrr) [inline src/infrastructure/collapsed_expression.rs:195]: This is unused. Is this a typo? Or supposed ot be `children`?

- [ ] check

File: src/infrastructure/collapsed_expression.rs [hunk 9]
```diff
@@ -130,9 +178,101 @@ impl Evaluation<CollapsedExpression<AbstractSyntaxTree>> for Node<'_> {
...
+                        match fun_name {
+                            Node::Leaf(Atom::Variable(Variable(var))) => {
+                                environment.insert(var.clone().into(), body);
+                            }
+                            dbg_info => {
+                                panic!("this should be anothe error {:?}", dbg_info)
+                            }
+                        }
+
+                        todo!()
```
Comment (pubrrr) [inline src/infrastructure/collapsed_expression.rs:220]: What should this even do? Is think an assignment should not even pop up here?

File: src/infrastructure/collapsed_expression.rs [hunk 10]
```diff
@@ -130,9 +178,101 @@ impl Evaluation<CollapsedExpression<AbstractSyntaxTree>> for Node<'_> {
...
+                    }
+                    (Node::Branch(op), Node::Branch(op2))
+                        if op2.operator() == Operator::Tuple
+                            && op.operator() == Operator::Tuple =>
+                    {
+                        let children1 = op.children();
+                        let children2 = op2.children();
+                        loop {
+                            let (head1, head2) = (
+                                children1.get(0).unwrap(),
```
Comment (pubrrr) [inline src/infrastructure/collapsed_expression.rs:230]: I think we should first design a proper interface before we mess with such things here...

**Review by pubrrr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**